### ISSUE TITLE
Issue running cypress tests on nx cloud

### ISSUE THREAD

**Issue by brandoncollins7** [nrwl/nx#20297]
### Current Behavior

When I run my CI cypress tests on the cloud I am getting a very weird crash:

![image](https://github.com/nrwl/nx/assets/36413482/592bf55f-f660-4628-b945-bd15b6b5c8a6)

### Expected Behavior

should run the tests the same as it does locally

### Steps to Reproduce

1. run cypress tests using nx cloud

### Nx Report

```shell
Node   : 18.15.0
   OS     : win32-x64
   pnpm   : 8.6.12

   nx (global)        : 17.0.1
   nx                 : 17.1.2
   @nx/js             : 17.1.2
   @nx/jest           : 17.1.2
   @nx/eslint         : 17.1.2
   @nx/workspace      : 17.1.2
   @nx/cypress        : 17.1.2
   @nx/devkit         : 17.1.2
   @nx/esbuild        : 17.1.2
   @nx/eslint-plugin  : 17.1.2
   @nx/nest           : 17.1.2
   @nx/next           : 17.1.2
   @nx/node           : 17.1.2
   @nx/react          : 17.1.2
   @nx/storybook      : 17.1.2
   @nrwl/tao          : 17.0.1
   @nx/vite           : 17.1.2
   @nx/web            : 17.1.2
   @nx/webpack        : 17.1.2
   typescript         : 5.1.6
   ---------------------------------------
   Community plugins:
   @nx-tools/nx-container : 5.0.2
   @nx-tools/nx-prisma    : 5.0.2
   ---------------------------------------
   The following packages should match the installed version of nx
     - @nrwl/tao@17.0.1
```

### Failure Logs

```shell
It looks like this is your first time using Cypress: 13.0.0

[STARTED] Task without title.
[FAILED] Xvfb exited with a non zero exit code.
[FAILED] 
[FAILED] There was a problem spawning Xvfb.
[FAILED] 
[FAILED] This is likely a problem with your system, permissions, or installation of Xvfb.
[FAILED] 
[FAILED] ----------
[FAILED] 
[FAILED] Error: (EE) 
[FAILED] Fatal server error:
[FAILED] (EE) Server is already active for display 99
[FAILED] If this server is no longer running, remove /tmp/.X99-lock
[FAILED] and start again.
[FAILED] (EE)
[FAILED] 
[FAILED] ----------
[FAILED] 
[FAILED] Platform: linux-x64 (Ubuntu - 22.04)
[FAILED] Cypress Version: 13.0.0
Xvfb exited with a non zero exit code.

There was a problem spawning Xvfb.

This is likely a problem with your system, permissions, or installation of Xvfb.

----------

Error: (EE) 
Fatal server error:
(EE) Server is already active for display 99
If this server is no longer running, remove /tmp/.X99-lock
and start again.
(EE)

----------

Platform: linux-x64 (Ubuntu - 22.04)
Cypress Version: 13.0.0
```

### Operating System

- [ ] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by AgentEnder** [comment]
Nx cloud doesn't currently run tasks on its own agents (see https://nx.dev/nx-cloud/intro/nx-cloud-workflows for upcoming solutions in that space), rather we just capture the outputs coming from your current CI provider. It seems that something in your pipeline definition is likely incorrect.

**Comment by antoinepairet** [comment]
- `Xvfb` is an in-memory display server for a UNIX-like operating system.
- It is used by Cypress in `headed` mode
- When running tests in parallel, spawning a single instance of `Xvfb` avoids conflicts; cf [Cypress docs](https://docs.cypress.io/guides/continuous-integration/introduction#Xvfb)

Cypress should not use `Xvfb` in headless mode. However, there were multiple issues with it in the past. For example:
- https://github.com/cypress-io/cypress/issues/16505

On CI, we run Storybook tests in parallel. Under the hood, it uses Cypress. To avoid issues, we ended up doing this ([Cypress docs](https://docs.cypress.io/guides/continuous-integration/introduction#Xvfb)):

````
          Xvfb :99 &
          export DISPLAY=:99

          EXECUTE YOUR TESTS IN // HERE

          pkill Xvfb
````

That being said, you probably do not want to run Cypress in `headed` mode (unless your tests require browser extensions?). 

If you want to run in `headed` mode, make sure that `Xvfb` is installed on your system. Here is how the official Cypress image is installing it: It is part of their official Docker image: https://github.com/cypress-io/cypress-docker-images/blob/master/base/18.12.1/Dockerfile

**Comment by brandoncollins7** [comment]
I am new to this so don't know anything about xvfg but I am just running cypress through nx as per the documentation.... So what could be wrong? Shouldn't the executor handle this properly by default?

"cypress-e2e": {
      "executor": "@nx/cypress:cypress",
      "options": {
        "cypressConfig": "cypress/tests/frnt-back-ui/cypress.config.ts",
        "devServerTarget": "frnt-back-ui:serve:cypress-e2e",
        "testingType": "e2e",
        "browser": "chrome",
        "env": {
          "NX_FRNT_BACKOFFICE_HOST": "http://localhost:3001"
        },
        "parallel": false
      },
      }
    },

**Comment by brandoncollins7** [comment]
my CI:

`name: Manual E2E

on:
  workflow_dispatch:

concurrency:
  group: cypress-e2e

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: ${{ github.head_ref }}
          token: ${{ secrets.GITHUB_TOKEN }}
          fetch-depth: 0

      - uses: pnpm/action-setup@v2
        with:
          version: 8.6.12

      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18.18.2
          cache: 'pnpm'

      - uses: nrwl/nx-set-shas@v3
        with:
          main-branch-name: ${{ github.base_ref }}
          workflow-id: 'e2e.yaml'

      - name: Cache Cypress binary
        uses: actions/cache@v2
        with:
          path: /home/runner/.cache/Cypress
          key: ${{ runner.os }}-cypress-${{ hashFiles('**/pnpm-lock.yaml') }}
          restore-keys: |
            ${{ runner.os }}-cypress-

      - name: Install dependencies
        run: |
          printf "\n //npm.staging.frnt.io/:_authToken="${{ secrets.NPM_AUTH_TOKEN }}"" >> .npmrc
          pnpm install --frozen-lockfile --prefer-offline

      - name: Install Cypress binary
        run: |
          if [ ! -d /home/runner/.cache/Cypress ]; then
            echo "Cypress cache not found. Installing Cypress."
            npx cypress install
          else
            echo "Cypress cache found. Skipping installation."
          fi

      - name: Initialize the Nx Cloud distributed CI run
        run: npx nx-cloud start-ci-run

      - name: Run commands in parallel
        env:
          INPUT_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          pids=()
          # list of commands to be run on main has env flag NX_CLOUD_DISTRIBUTED_EXECUTION set to false
          #NX_CLOUD_DISTRIBUTED_EXECUTION=false VERSION=${{ github.sha }} ENVIRONMENT=uat npx nx affected --base=\$NX_BASE --head=\$NX_HEAD -t deploy --parallel=3 --nx-bail=true & pids+=($!)

          npx nx run-many --target=cypress-e2e --configuration=ci --all --record --key 46362436-0e3b-48e0-a977-c92e35d88b16 &
          pids+=($!)
          echo "PID for nx container: ${pids[-1]}"

          # run all commands in parallel and bail if one of them fails
          for pid in "${pids[@]}"; do
            echo "Waiting for pid $pid"
            if ! wait $pid; then
              echo "Process $pid exited with status $?"
              exit 1
            fi
          done

          exit 0

      - name: Stop all running agents for this CI run
        # It's important that we always run this step, otherwise in the case of any failures in preceding non-Nx steps, the agents will keep running and waste billable minutes
        if: ${{ always() }}
        run: npx nx-cloud stop-all-agents

  agents:
    name: Agent ${{ matrix.agent }}
    runs-on: ubuntu-latest

    strategy:
      matrix:
        agent: [ 1, 2, 3 ]
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: ${{ github.head_ref }}
          token: ${{ secrets.GITHUB_TOKEN }}
          fetch-depth: 0

      - uses: pnpm/action-setup@v2
        with:
          version: 8.6.12

      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18.18.2
          cache: 'pnpm'

      - uses: nrwl/nx-set-shas@v3
        with:
          main-branch-name: ${{ github.base_ref }}
          workflow-id: 'e2e.yaml'

      - name: Cache Cypress binary
        uses: actions/cache@v2
        with:
          path: /home/runner/.cache/Cypress
          key: ${{ runner.os }}-cypress-${{ hashFiles('**/pnpm-lock.yaml') }}
          restore-keys: |
            ${{ runner.os }}-cypress-

      - name: Install dependencies
        run: |
          printf "\n //npm.staging.frnt.io/:_authToken="${{ secrets.NPM_AUTH_TOKEN }}"" >> .npmrc
          pnpm install --frozen-lockfile --prefer-offline

      - name: Install Cypress binary
        run: |
          if [ ! -d /home/runner/.cache/Cypress ]; then
            echo "Cypress cache not found. Installing Cypress."
            npx cypress install
          else
            echo "Cypress cache found. Skipping installation."
          fi

      - name: Start Nx Agent ${{ matrix.agent }}
        run: npx nx-cloud start-agent
        env:
          NX_AGENT_NAME: ${{matrix.agent}}
          INPUT_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
`

**Comment by FrozenPandaz** [comment]
A lot has changed in our [Nx Cloud documentation regarding Nx Agents](https://nx.dev/ci/features/dynamic-agents#dynamically-allocate-agents). Please read that guide and see if it works for you.

I'm going to close this issue as I believe a lot has changed. If you still have issues, please open a new issue with more updated information.
